### PR TITLE
docs: QLoRA Documentation and Notebooks

### PR DESCRIPTION
**Pull request by RexBearIU**
# Description

 This PR adds comprehensive documentation and tutorials for running LoRA/QLoRA fine-tuning, specifically focusing on multi-host TPU environments.

  As PEFT techniques become more prevalent for large models, users need clear, step-by-step guidance on how to leverage MaxText and Tunix for multi-host tuning.

  This PR includes:
   - A new dedicated Markdown tutorial (docs/tutorials/posttraining/lora_on_multi_host.md) detailing the prerequisites, environment setup, and execution steps for multi-host TPU tuning.
   - A fully documented Jupyter Notebook (src/maxtext/examples/lora_llama3_demo.ipynb) that walks through setup, pre-SFT evaluation (using vLLM rollouts), the Tunix LoRA SFT training loop, and post-SFT evaluation with weight
     merging.
   - Minor path corrections in the existing lora.md tutorial.

# Tests

  - Documentation only.
  - The Jupyter notebook has been run manually to ensure that all cells execute correctly (assuming the required dependencies and TPUs are available).

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [X] I have performed a self-review of my code. For an optional AI review, add the `gemini-review` label.
- [X] I have necessary comments in my code, particularly in hard-to-understand areas.
- [X] I have run end-to-end tests tests and provided workload links above if applicable.
- [X] I have made or will make corresponding changes to the doc if needed, including adding new documentation pages to the relevant Table of Contents (toctree directive) as explained in [our documentation](https://maxtext.readthedocs.io/en/latest/development.html#adding-new-documentation-files).

### CODE REVIEW

File: docs/tutorials/posttraining/lora_on_multi_host.md
Comment (SurbhiJainUSC) [inline docs/tutorials/posttraining/lora_on_multi_host.md:1]: Add this here: https://github.com/AI-Hypercomputer/maxtext/blob/main/docs/tutorials/post_training_index.md?plain=1#L30

File: docs/tutorials/posttraining/lora_on_multi_host.md [hunk 1]
```diff
@@ -0,0 +1,304 @@
...
+- Access to a Google Cloud Project with TPU quotas.
+- A Hugging Face account with an access token for downloading models.
+- Permissions for Google Artifact Registry (Artifact Registry Writer role).
+- Prerequisites for XPK installed (follow [official documentation](https://github.com/AI-Hypercomputer/xpk/blob/main/docs/installation.md#1-prerequisites)).
+- A Pathways-ready GKE cluster (see [create GKE cluster](https://docs.cloud.google.com/ai-hypercomputer/docs/workloads/pathways-on-cloud/create-gke-cluster)).
+- **Docker** installed and configured for sudoless use. Follow the steps to [configure sudoless Docker](https://docs.docker.com/engine/install/linux-postinstall/).
+
+## Build and upload MaxText Docker image
+
+For instructions on building and uploading the MaxText Docker image with post-training dependencies, please refer to the [official documentation](https://maxtext.readthedocs.io/en/latest/build_maxtext.html).
```
Comment (SurbhiJainUSC) [inline docs/tutorials/posttraining/lora_on_multi_host.md]: Change absolute path of `build_maxtext.html` to relative path to `build_maxtext.md`

File: docs/tutorials/posttraining/lora_on_multi_host.md [hunk 2]
```diff
@@ -0,0 +1,304 @@
...
+
+## Create GKE cluster
```
Comment (SurbhiJainUSC) [inline docs/tutorials/posttraining/lora_on_multi_host.md]: This can be removed as we already have this mentioned in `Prerequisites`.

File: docs/tutorials/posttraining/lora_on_multi_host.md [hunk 3]
```diff
@@ -0,0 +1,304 @@
...
+
+### Option 1: Using an existing MaxText checkpoint
+
+If you already have a MaxText-compatible model checkpoint, simply set the following environment variable and move on to the next section.
+
+```bash
+export MAXTEXT_CKPT_PATH=<CKPT_PATH> # e.g., gs://my-bucket/my-model-checkpoint/0/items
+```
+
+**Note:** Make sure that `MAXTEXT_CKPT_PATH` has the checkpoints created using the correct storage flags:
```
Comment (SurbhiJainUSC) [inline docs/tutorials/posttraining/lora_on_multi_host.md:143]: Update line 147-153 as:
```
**Note:** Make sure that `MAXTEXT_CKPT_PATH` has the checkpoints created using the correct storage flags:

```bash
checkpoint_storage_use_zarr3=False
checkpoint_storage_use_ocdbt=False
```
```

File: docs/tutorials/posttraining/lora_on_multi_host.md [hunk 4]
```diff
@@ -0,0 +1,304 @@
...
+export MAXTEXT_CKPT_PATH=<CKPT_PATH> # gs://my-bucket/my-checkpoint-directory/0/items
+```
+
+## Submit workload on GKE cluster
+
+This section provides the command to run LoRA Fine-Tuning on a GKE cluster.
+
+### Run a Fresh LoRA Fine-Tuning on Hugging Face Dataset
+
+#### LoRA with Multi-Controller JAX (McJAX)
```
Comment (SurbhiJainUSC) [inline docs/tutorials/posttraining/lora_on_multi_host.md]: Remove McJAX section

File: docs/tutorials/posttraining/lora_on_multi_host.md [hunk 5]
```diff
@@ -0,0 +1,304 @@
...
+
+xpk workload create-pathways \
+--cluster=${GKE_CLUSTER?} \
+--project=${PROJECT_ID?} \
+--zone=${ZONE?} \
+--docker-image=${DOCKER_IMAGE?} \
+--workload=${RUN_NAME?} \
+--tpu-type=${TPU_TYPE?} \
+--num-slices=${NUM_SLICES?} \
+--command="JAX_PLATFORMS=proxy JAX_BACKEND_TARGET=grpc://127.0.0.1:29000 ENABLE_PATHWAYS_PERSISTENCE=1 python3 -m maxtext.trainers.post_train.sft.train_sft run_name=${RUN_NAME?} base_output_directory=${BASE_OUTPUT_DIRECTORY?} model_name=${MODEL?} load_parameters_path=${MAXTEXT_CKPT_PATH?} hf_access_token=${HF_TOKEN?} hf_path=${DATASET_NAME?} train_split=${TRAIN_SPLIT?} hf_data_dir=${HF_DATA_DIR?} train_data_columns=${TRAIN_DATA_COLUMNS?} steps=${STEPS?} per_device_batch_size=${PER_DEVICE_BATCH_SIZE?} max_target_length=${MAX_TARGET_LENGTH?} learning_rate=${LEARNING_RATE?} chat_template_path=${CHAT_TEMPLATE_PATH?} enable_nnx=True pure_nnx_decoder=True lora.enable_lora=True lora.lora_rank=${LORA_RANK?} lora.lora_alpha=${LORA_ALPHA?} checkpoint_storage_use_zarr3=$((1 - USE_PATHWAYS)) checkpoint_storage_use_ocdbt=$((1 - USE_PATHWAYS)) enable_single_controller=True"
```
Comment (SurbhiJainUSC) [inline docs/tutorials/posttraining/lora_on_multi_host.md]: Break the command in multiple lines for readability

File: src/maxtext/examples/lora_llama3_demo.ipynb [hunk 6]
```diff
@@ -0,0 +1,610 @@
...
+    "    print(\n",
+    "        f\"Percentage of test samples where the model's output adheres to the expected structure: {score['correct_format']}%\"\n",
+    "    )\n",
+    "else:\n",
+    "    print(\"Evaluation on test dataset is skipped. Please set `RUN_EVALUATION=True` to run evaluation.\")"
+   ]
+  }
+ ],
+ "metadata": {
+  "kernelspec": {
```
Comment (SurbhiJainUSC) [inline src/maxtext/examples/lora_llama3_demo.ipynb:590]: nit:
```
"kernelspec": {
   "display_name": "Python (3.12.11)",
   "language": "python",
   "name": "python3"
  },
```

File: docs/tutorials/posttraining/lora_on_multi_host.md
Comment (RexBearIU) [reply docs/tutorials/posttraining/lora_on_multi_host.md:1]: Added! I have included a link to this new tutorial under the LoRA section in  `docs/tutorials/post_training_index.md`

File: docs/tutorials/posttraining/lora_on_multi_host.md (hunk 1 above)
Comment (RexBearIU) [reply docs/tutorials/posttraining/lora_on_multi_host.md]: Switched the absolute link of  `build_maxtext.html`  to the relative path  `../../build_maxtext.md`

File: docs/tutorials/posttraining/lora_on_multi_host.md (hunk 2 above)
Comment (RexBearIU) [reply docs/tutorials/posttraining/lora_on_multi_host.md]: Good point, I have removed this redundant section entirely since it's already covered in the Prerequisites.

File: docs/tutorials/posttraining/lora_on_multi_host.md (hunk 3 above)
Comment (RexBearIU) [reply docs/tutorials/posttraining/lora_on_multi_host.md:143]: Updated this section to use the clean boolean ` False ` settings as recommended.

File: docs/tutorials/posttraining/lora_on_multi_host.md (hunk 4 above)
Comment (RexBearIU) [reply docs/tutorials/posttraining/lora_on_multi_host.md]: Cleaned this up and removed all McJAX sections and headers, focusing the tutorial purely on the Pathways path.

File: docs/tutorials/posttraining/lora_on_multi_host.md (hunk 5 above)
Comment (RexBearIU) [reply docs/tutorials/posttraining/lora_on_multi_host.md]: Broken down both  `xpk`  commands (fresh and restore) across multiple lines for better readability.

File: src/maxtext/examples/lora_llama3_demo.ipynb (hunk 6 above)
Comment (RexBearIU) [reply src/maxtext/examples/lora_llama3_demo.ipynb:590]: Nice catch, updated the notebook metadata to use the standard  python3  kernelspec.

File: docs/guides/lora_model_bringup.md [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+
+## 1. Step-by-Step Bring-up Guide for NNX LoRA
+
+To enable LoRA support for a new model, follow these two simple steps:
+
+### Step 1.1: Verify Base Model Support
+
+The target model architecture must already be implemented and supported as a base model in MaxText.
+
+- The JAX/NNX model definition should be located under `src/maxtext/models/` (e.g., \[gemma3.py\](file:///home/jackyf_google_com/maxtext/src/maxtext/models/gemma3.py)).
```
Comment (igorts-git) [inline docs/guides/lora_model_bringup.md]: Please check the links in this doc: `file:///home/jackyf_google_com/...`

File: docs/tutorials/posttraining/lora_on_multi_host.md [hunk 8]
```diff
@@ -0,0 +1,310 @@
...
+    model_name=${MODEL?} \
+    lora.lora_restore_path=${BASE_OUTPUT_DIRECTORY?}/${RUN_NAME?}/checkpoints/<STEPS>/model_params \
+    base_output_directory=${BASE_OUTPUT_DIRECTORY?}/hf_lora_adapter \
+    hf_access_token=${HF_TOKEN?}"
+    
+```
+
+- `lora.lora_restore_path`: Point this to the specific checkpoint directory (e.g., `.../checkpoints/1000/items`) that you want to export.
+- `base_output_directory`: The local or GCS directory where the Hugging Face `adapter_model.safetensors` and `adapter_config.json` will be saved.
+- `lora.lora_rank` / `lora.lora_alpha`: Must match the values used during the training phase to ensure the `adapter_config.json` is generated correctly.
```
Comment (igorts-git) [inline docs/tutorials/posttraining/lora_on_multi_host.md:305]: I am wondering whether we could save these values in the MaxText checkpoint, such that the users won't have to provide them here again and potentially make a mistake?

This is just a question and an idea for a future improvement. We don't need to block the quantization rollout on this.

File: docs/guides/lora_model_bringup.md [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+
+If no prefix matches your model name, MaxText falls back to the `default` pattern:
+
+```yaml
+default: "decoder/layers/.*(self_attention/(query|key|value|out)|mlp/(wi_0|wi_1|wo))"
+```
+
+______________________________________________________________________
+
+## 2. Integrating Custom Weight Mappings (When is it needed?)
```
Comment (igorts-git) [inline docs/guides/lora_model_bringup.md:55]: If a user follows these instructions they would not be very confident that they succeeded. We need to add a section here that describes how to check which layers were matched by my regex and which layers did not.

File: docs/guides/lora_model_bringup.md (hunk 7 above)
Comment (RexBearIU) [reply docs/guides/lora_model_bringup.md]: Oops, thanks for catching that! The absolute local file URIs were accidentally left in. 

I've updated all occurrences of those absolute links in `docs/guides/lora_model_bringup.md` to be relative paths (e.g., pointing to `../../src/maxtext/models/gemma3.py`). These fixes are included in the latest push.

File: docs/tutorials/posttraining/lora_on_multi_host.md (hunk 8 above)
Comment (RexBearIU) [reply docs/tutorials/posttraining/lora_on_multi_host.md:305]: That is an excellent suggestion! 

I have addressed this in a dedicated PR: **#4269** (`jackyf/lora-ckpt-metadata`). 

This PR implements an automated `lora_config.json` metadata sidecar writer inside `save_checkpoint` of `checkpointing.py` and syncs/loads it inside `to_huggingface.py` so that users don't have to specify rank/alpha manually anymore.

File: docs/guides/lora_model_bringup.md (hunk 9 above)
Comment (RexBearIU) [reply docs/guides/lora_model_bringup.md:55]: Agreed, that's a very important validation step.

To provide a clear validation path:
1. **Interactive Verification**: Added a new section **"3. Verifying Your Custom LoRA Targets"** to this PR, providing a lightweight Python snippet utilizing `nnx.iter_modules` to inspect matched layers programmatically.
2. **Proper Startup Logging (PR #4067)**: Implemented automatic log-based verification in `lora_utils.py` as part of the refactor PR #4067. Once merged, MaxText will traverse the graph via `nnx.iter_graph` on startup, log the count of matched submodules, print a preview sample of wrapped layers, or fail loudly if zero modules match.

File: docs/guides/lora_model_bringup.md [hunk 10]
```diff
@@ -0,0 +1,113 @@
...
+- The conversion utility automatically maps, scales, and formats the LoRA adapter parameters back into standard Hugging Face PEFT format based on the base model's existing weight mapping.
+
+### Scenario B: Decoding with the MaxText vLLM Adapter (Mapping is Required)
+
+If you want to perform decoding or run high-performance serving on your adapted model using the **MaxText vLLM adapter** (e.g., via `vllm_decode`), **you must define and register a custom weight mapping.** This allows the vLLM JAX wrapper to dynamically map and feed weights to the vLLM engine.
+
+To add weight mapping for vLLM decode:
+
+1. **Create a Weight Mapping Config**:
+   Create a new file in \[src/maxtext/integration/tunix/weight_mapping/\](../../src/maxtext/integration/tunix/weight_mapping/) (e.g., `your_model.py`) defining a mapping dataclass. You can refer to \[gemma3.py\](../../src/maxtext/integration/tunix/weight_mapping/gemma3.py) or \[llama3.py\](../../src/maxtext/integration/tunix/weight_mapping/llama3.py) as templates.
```
Comment (igorts-git) [inline docs/guides/lora_model_bringup.md:72]: Our documentation is usually surfaced via https://maxtext.readthedocs.io/en/latest/index.html I am not sure that the hyperlinks to code in GitHub would work here. Please check how code links are implemented in other docs.  We are also trying to keep docs consistent w.r.t. MaxText release versions. Such that if someone is reading docs for version 0.2.3, all hyperlinks also point to the same version.

@melissawm do you know how to correctly link to code in GitHub?

Comment (melissawm) [reply docs/guides/lora_model_bringup.md:72]: Yes, these will not work. The readthedocs site can only see relative links to documents under the docs/ folder. To link to files under src/ or other folders, the best way is to use the github link (in this case, `https://github.com/AI-Hypercomputer/maxtext/blob/main/src/maxtext/integration/tunix/weight_mapping/llama3.py`)

File: docs/guides/lora_model_bringup.md (hunk 9 above)
Comment (igorts-git) [reply docs/guides/lora_model_bringup.md:55]: OK, thanks

**Review by igorts-git — APPROVED** [review]

**Review by shralex — APPROVED** [review]

File: src/maxtext/examples/lora_llama3_demo.ipynb [hunk 11]
```diff
@@ -0,0 +1,610 @@
...
+   "outputs": [],
+   "source": [
+    "if IN_COLAB:\n",
+    "    # Clone the MaxText repository\n",
+    "    !git clone https://github.com/AI-Hypercomputer/maxtext.git\n",
+    "    %cd maxtext\n",
+    "\n",
+    "    # Install uv, a fast Python package installer\n",
+    "    !pip install uv\n",
+    "    \n",
```
Comment (SurbhiJainUSC) [inline src/maxtext/examples/lora_llama3_demo.ipynb:81]: Please add this:
```
import os
os.environ["UV_TORCH_BACKEND"]="cpu"
    
```

File: src/maxtext/examples/lora_llama3_demo.ipynb [hunk 12]
```diff
@@ -0,0 +1,610 @@
...
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "cf6cdced",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "if not epath.Path(MODEL_CHECKPOINT_PATH).exists():\n",
+    "    # Install torch for the conversion script\n",
+    "    print(\"Installing torch...\")\n",
+    "    subprocess.run(\n",
```
Comment (SurbhiJainUSC) [inline src/maxtext/examples/lora_llama3_demo.ipynb:220]: I don't think we need to install `torch` for checkpoint conversion script anymore.

**Comment by RexBearIU** [comment]
Hi @SurbhiJainUSC  for the remaining issues you bring up I will modify in another PR and let you know.

**Comment by RexBearIU** [comment]
I have created a new PR #4277 to address the leftover comments in the `lora_llama3_demo.ipynb` notebook.

**Comment by RexBearIU** [comment]
I have also updated the branch in #4277 to resolve the ReadTheDocs relative links issue in `docs/guides/lora_model_bringup.md` as discussed by @melissawm. All relative links to code files under `src/` are now proper GitHub absolute URLs.
